enrollment-server/src/main/java/com/wultra/app/enrollmentserver/impl/service/converter/MobileTokenConverter.java

```diff
@@ -95,46 +101,61 @@ public Operation convert(OperationDetailResponse operationDetail, OperationTempl
+            formData.setMessage(sub.replace(operationTemplate.getMessage()));
+        } else {
+            formData.setTitle(operationTemplate.getTitle());
+            formData.setMessage(operationTemplate.getMessage());
+        }
 
-            final String attributes = operationTemplate.getAttributes();
-            if (attributes != null) {
-                final OperationTemplateParam[] operationTemplateParams = objectMapper.readValue(attributes, OperationTemplateParam[].class);
-                if (operationTemplateParams != null) {
-                    for (OperationTemplateParam templateParam : operationTemplateParams) {
-                        final Attribute attribute = buildAttribute(templateParam, parameters);
-                        if (attribute != null) {
-                            formData.getAttributes().add(attribute);
-                        }
+        final String attributes = operationTemplate.getAttributes();
+        if (attributes != null) {
+            final OperationTemplateParam[] operationTemplateParams = objectMapper.readValue(attributes, OperationTemplateParam[].class);
+            if (operationTemplateParams != null) {
+                for (OperationTemplateParam templateParam : operationTemplateParams) {
+                    final Attribute attribute = buildAttribute(templateParam, parameters);
+                    if (attribute != null) {
+                        formData.getAttributes().add(attribute);
                     }
                 }
             }
+        }
+        return formData;
+    }
 
-            operation.setFormData(formData);
-            return operation;
-        } catch (JsonProcessingException e) {
-            logger.debug("Unable to parse JSON with operation template parameters: {}", e.getMessage());
-            logger.debug("Exception detail", e);
-            throw new MobileTokenConfigurationException("ERR_CONFIG", "Invalid JSON structure for the configuration: " + e.getMessage());
+    private static StringSubstitutor createStringSubstitutor(final Map<String, String> parameters) {
+        if (CollectionUtils.isEmpty(parameters)) {
+            return null;
+        } else {
+            return new StringSubstitutor(parameters);
         }
     }
 
-    private UiExtensions convertUiExtension(final OperationDetailResponse operationDetail, final OperationTemplateEntity operationTemplate) throws JsonProcessingException {
+    private UiExtensions convertUiExtension(
+            final OperationDetailResponse operationDetail,
+            final OperationTemplateEntity operationTemplate,
+            final StringSubstitutor substitutor) throws JsonProcessingException {
+
         if (StringUtils.hasText(operationTemplate.getUi())) {
-            final String uiJsonString = operationTemplate.getUi();
+            final String uiJsonString = substitutor == null ? operationTemplate.getUi() : substitutor.replace(operationTemplate.getUi());
```

Comment:
It is possible to substitute the whole UI.

Reply:
I am not sure this is a good approach.

Reply:
As far as I remember, that was an agreement from the last meeting, but unfortunately is not written down :(

Reply:
My suggestion is to allow substitution of the payload for the `GENERIC` subtype. AFAIK there is no other need for substitution at the moment, but we can of course discuss it. @kober32

Reply:
As we agreed on the call, the whole UI will be substituted. 

- [ ] Add test case for dangerous json characters